README.md
# flutter_guid

A flutter package that emulates the C# type Guid. Use this package to:
1. Generate new UUIDs.
1. Add variables of type Guid to your classes.
1. Compare two 'Guid' variables safely against each other.
1. Guarantee validity of data when working with UUID values.

## Getting Started
### Installing
Install from https://pub.dev/packages/flutter_guid#-installing-tab- , "flutter_guid's installation page on pub.dev"

### Usage
#### Add to your class.
```dart
import 'package:flutter_guid/flutter_guid.dart';

class Employee{
  final Guid id;
  final Guid departmentId;

  Employee({this.id, this.departmentId});
}
```

#### Initialize Guid values
```dart

// initialize a guid from a string
final accountsDepartmentId = new Guid("41e879aa-859c-4bda-b77b-37b7e07c6674");

 // Generate a guid on the fly (Uses the superb package 'https://pub.dev/packages/uuid').
final employeeId = Guid.newGuid;

```

#### Compare Guids
```dart
// initialize a guid from a lowercase string
final accountsDepartmentIdLowerCase = new Guid("41e879aa-859c-4bda-b77b-37b7e07c6674");

// initialize a guid with same string, but UPPERCASE
final accountsDepartmentIdUpperCase = new Guid("41E879AA-859C-4BDA-B77B-37B7E07C6674");

// performs case-insensitive comparison and evaluates to 'true'
final equal = accountsDepartmentIdUpperCase == accountsDepartmentIdLowerCase;

```

#### Getting the UUID value contained in the GUID
```dart
final Guid accountsDepartmentIdLowerCase = new Guid("41e879aa-859c-4bda-b77b-37b7e07c6674");

// this works.
String uuidString = accountsDepartmentIdLowerCase.value;

// this also works.
uuidString = accountsDepartmentIdLowerCase.toString();
```

## Testing
### Basic Testing
Clone the repo and run ```flutter test```
### Testing With Coverage Information (Linux only)
1. Ensure you have lcov installed and setup on your system (http://ltp.sourceforge.net/coverage/lcov.php).
2. Ensure the run-tests script in the project root is executable (```chmod +x ./run-tests.sh```).
3. Execute the run-tests script in the project root ```./run-tests.sh```.
4. Navigate into directory ./coverage/html/ (relative to project root). 
5. Open index.html to view coverage information.

### Coverage Information On Other Systems
It is unlikely that I'll be porting the run-tests.sh script to other systems, my reason mostly being that I don't frequently use any other OS and would not want to pick a sub-optimal tool for coverage generation (also laziness probably is a factor).

Anyhow, it should be fairly straight-forward to do the porting if you so wish.

The run-tests.sh consists of all of two lines:
1. flutter test --coverage
1. genhtml ./coverage/lcov.info  -o ./coverage/html

The first line is platform independent and generates the file ./coverage/lcov.info (again relative to project root).
On Linux, we can consume the coverage info contained in this file via the lcov tool. On Windows and other
platforms, there exists also tools that would be able to similarly consume the coverage information and produce human friendly
output.

In essense, port the script to desired platform by
1. Keeping the first line 'as is'.
1. Finding a tool to process the ./coverage/lcov.info file.
1. Profit!


CHANGELOG.md
## [0.2.0] - Added Null Safety

* Added tests and information in Readme.md
## [0.1.0] - Documentation and Testing

* Added tests and information in Readme.md

## [0.0.1] - Initial Release.

* Describe initial release.




test/guid_test.dart
import 'package:flutter_guid/flutter_guid_error.dart';
import 'package:flutter_guid/flutter_guid.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group("Equality Tests:", () {
    test("Case insensitive equality works", () {
      final guidAlpha = new Guid("ac3fa00f-8f5b-4e93-b7a5-2ee3051a12b9");
      final guidBeta = new Guid("AC3fa00f-8f5b-4e93-b7a5-2ee3051a12b9");

      final areEqual = guidAlpha == guidBeta;

      expect(true, areEqual);
    });

    test("Null values in testing don't throw exception", () {
      final guidAlpha = new Guid("");
      Guid? guidBeta;

      final areEqual = guidAlpha == guidBeta;

      expect(areEqual, false);
    });
  });

  group("Validity Tests", () {
    _initializer(String? value, bool expectException) {
      if (expectException) {
        expect(() => new Guid(value), throwsA((e) {
          return e != null && e is FlutterGuidError;
        }));
      } else {
        final guid = new Guid(value);
        final expectedValue =
            (value == null || value.isEmpty) ? Guid.defaultValue.value : value;
        expect(guid.value, expectedValue);
        expect(guid.toString(), expectedValue);
      }
    }

    test("Invalid decimal values not accepted",
        () => _initializer("12345", true));
    test("Invalid hex values not accepted", () => _initializer("bad", true));

    test("Invalid random values not accepted",
        () => _initializer("thequickbrown", true));

    test("Invalid decimal and hex values not accepted",
        () => _initializer("12345", true));

    test("Default Guid is allowed",
        () => _initializer("00000000-0000-0000-0000-000000000000", false));

    test("Valid Guid is allowed",
        () => _initializer("f841928c-5393-4c6e-9b22-85de1fcf317a", false));

    test("Null is allowed", () => _initializer(null, false));

    test("Empty string is allowed", () => _initializer("", false));

    test("Get newGuid works", () {
      final guid = Guid.newGuid;
      final isValid = Guid.isValid(guid);
      expect(isValid, true);
    });
  });
}


lib/flutter_guid.dart
library flutter_guid;

import 'package:flutter_guid/flutter_guid_error.dart';
import 'package:uuid/uuid.dart';
import 'package:validators/validators.dart';

/// Class that emulates as closely as possible the C# Guid type.
class Guid {
  static const String _defaultGuid = "00000000-0000-0000-0000-000000000000";

  /// The Guid whose value is the default sequence of characters that represent a 'zero-value' UUID in .Net "00000000-0000-0000-0000-000000000000"
  static Guid get defaultValue => new Guid(_defaultGuid);

  String? _value;

  /// Generates a new v4 UUID during instantiation and returns a GUID instance with the new UUID.
  Guid.generate() : this(Guid.newGuid.value);

  /// Constructor, expects a valid UUID and will throw an exception if the value provided is invalid.
  Guid(String? v) {
    _failIfNotValidGuid(v);
    _value = v;
  }

  /// Generates a new v4 UUID and returns a GUID instance with the new UUID.
  static Guid get newGuid {
    return new Guid(Uuid().v4());
  }

  /// Checks whether a value is a valid Guid
  /// Returns false if 'guid' is null or has an invalid value
  /// Returns true if guid is valid
  static bool isValid(Guid? guid) {
    if (guid == null) {
      return false;
    } else {
      return isUUID(guid.value);
    }
  }

  _failIfNotValidGuid(String? v) {
    if (v == null || v.isEmpty) {
      v = _defaultGuid;
    }
    final isInvalid = isUUID(v) == false;
    if (isInvalid) {
      throw new FlutterGuidError("Value '$v' is not a valid UUID");
    }
  }

  /// Gets the UUID value contained by the Guid object as a string.
  String get value {
    if (_value == null || _value!.isEmpty) {
      return _defaultGuid;
    } else {
      return _value!;
    }
  }

  /// Performs a case insensitive comparison on the UUIDs contained in two Guid objects.
  /// Comparison is by value and not by reference.
  bool operator ==(other) {
    return this.value.toLowerCase() == other.toString().toLowerCase();
  }

  /// Returns the hashCode.
  @override
  int get hashCode {
    return super.hashCode;
  }

  /// Gets the UUID value contained by the Guid object as a string.
  @override
  String toString() {
    return value;
  }
}


example/example.dart
import 'package:flutter_guid/flutter_guid.dart';

import 'employee.dart';

class EmployeeManager {
  static final String workShopGuidString =
      "edea6638-dd1d-4e26-a207-e4044ad22687";
  final employeeList = <Employee>[];

  // initialize a guid from a string
  final accountsDepartmentId = new Guid("41e879aa-859c-4bda-b77b-37b7e07c6674");

  // initialize a new guid with the 'workShopGuidString' value, Lowercased.
  final workshopDepartmentIdLowerCased =
      new Guid(workShopGuidString.toLowerCase());

  // initialize a second guid with the 'workShopGuidString' value, BUT UPPERCASED this time.
  final workshopDepartmentIdUpperCased =
      new Guid(workShopGuidString.toUpperCase());

  populateEmployeeList() {
    final accountant = new Employee(
        departmentId: accountsDepartmentId, // use pre-initialized guid value.
        id: Guid
            .newGuid // Generate a guid on the fly (Uses the superb package 'https://pub.dev/packages/uuid').
        );

    final engineer = new Employee(
        departmentId: workshopDepartmentIdLowerCased, id: Guid.newGuid);

    employeeList
      ..add(accountant)
      ..add(engineer);
  }

  Employee getEngineer() {
    final theEngineer = employeeList.singleWhere((employee) {
      //Comparison done on guid value is case-insensitive
      //employee.departmentId was initialized with a lowercase string but is compared to a guid initialized with an uppercase string and WILL SUCCESSFULLY match
      return employee.departmentId == workshopDepartmentIdUpperCased;
    });

    return theEngineer;
  }

  bool engineerIdIsDefault() {
    final engineer = getEngineer();

    // Returns true if engineerId == "00000000-0000-0000-0000-000000000000".
    // This value is assigned to Guids if no other value is explicitly set on them
    // NOTE: The guid variable itself if nullable but, its value once initialized is guaranteed to always be either the defaultValue or a valid UUID
    return engineer.id == Guid.defaultValue;
  }
}
